### PR TITLE
BCC upstream_submit (temporary; for discussion only) 

### PR DESCRIPTION
**Pull request by jcanseco**
## About this PR
This PR is to serve as _the_ place where we discuss the patches we wish to submit to BCC upstream. Since this PR is for discussion only, this will never actually be approved and merged. 

The intention is to start small and cherry-pick only a few commits from the _joelagnel/bcc/bcc-bpfd_ branch onto **_jcanseco/bcc/upstream_submit_** (which was branched off _iovisor/bcc/master_ and is thus up-to-date with "upstream").

From the discussion, I will make the appropriate changes, whether it be code changes or commit squashing/splitting/rewording.

Once we finish discussing the commits and are satisfied, we can continue cherry-picking more commits from _joelagnel/bcc/bcc-bpfd_ and repeat the discussion once again. 

This is therefore an iterative, ongoing process, and this PR serves as the platform where we can coordinate and document our discussions during this said process.

## Testing
We can test our upstream submission for BCC by running the BCC tools against HiKey / ShellRemote via BPFd.

However, there's a catch: since BPFd depends on BCC APIs that have been changed upstream, this requires changes to BPFd itself. The BPFd source files in _jcanseco/**bcc**/upstream_submit_ therefore need to deviate from those in _joelagnel/**bpfd**/master_.

This won't do since we want to test our BCC upstream submission against the same BPFd code we intend to submit upstream, but as it is currently, only the BPFd repository has the CMake configs required to build BPFd for arm64 (which we need if we wish to test against HiKey). 

So, we introduce another branch: _jcanseco/**bpfd**/upstream_submit_, which is a branch in the BPFd repo whose source files are kept in sync with _jcanseco/**bcc**/upstream_submit_. This allows us to build BPFd for arm64 from the same sources we intend to upstream, which we can then use to test our BCC upstream submission.

tl;dr: For testing, we run the tools against the bpfd executable built from _jcanseco/**bpfd**/upstream_submit_. All comments regarding tests that have been run for our BCC upstream submission will be about tests run with this set-up.

## Notes
- GitHub [does not show the commits in the right order](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/) when the history has been modified using rebase / force push (which we use liberally to massage the structure of our upstream submission). If you'd like to see the right order, please checkout this branch locally and invoke `git log`.

### CODE REVIEW

File: src/cc/frontends/clang/CMakeLists.txt [hunk 1]
```diff
@@ -10,3 +10,4 @@ if(DEFINED BCC_BACKUP_COMPILE)
 endif()
 
 add_library(clang_frontend STATIC loader.cc b_frontend_action.cc tp_frontend_action.cc kbuild_helper.cc ../../common.cc)
+target_link_libraries(clang_frontend bcc-static)
```
Comment (joelagnel) [inline src/cc/frontends/clang/CMakeLists.txt:13]: Drop this commit. Its already merged upstream..

File: src/cc/bpfd/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) Jazel Canseco, 2018
+# Licensed under the Apache License, Version 2.0 (the "License")
+
+add_executable(bpfd bpfd.c base64.c remote_perf_reader.c utils.c)
+target_link_libraries(bpfd bpf-static)
```
Comment (joelagnel) [inline src/cc/bpfd/CMakeLists.txt]: I was thinking for the upstream port, we really don't want static, it will raise a red flag.. Can you change it to be dynamic linked for upstream?

**Review by joelagnel — CHANGES_REQUESTED** [review]

File: src/cc/bpfd/CMakeLists.txt (hunk 2 above)
Comment (joelagnel) [reply src/cc/bpfd/CMakeLists.txt]: What would be perfect is if you can make upstream dynamic, and put the static stuff in the arm build.. so this master file has nothing static. You can use a flag for that which will be set by the ARM specific compile cmake file

File: src/cc/frontends/clang/CMakeLists.txt (hunk 1 above)
Comment (jcanseco) [reply src/cc/frontends/clang/CMakeLists.txt:13]: It was removed in commit [29015d7ef003f35f2a4671d67404f9e5647d6b13](https://github.com/iovisor/bcc/commit/29015d7ef003f35f2a4671d67404f9e5647d6b13)

File: src/cc/bpfd/CMakeLists.txt (hunk 2 above)
Comment (jcanseco) [reply src/cc/bpfd/CMakeLists.txt]: Done.

File: src/cc/frontends/clang/CMakeLists.txt (hunk 1 above)
Comment (jcanseco) [reply src/cc/frontends/clang/CMakeLists.txt:13]: Okay so I just looked into this issue. Commit 29015d7  removed the line linking `clang_frontend` to `bcc_static` because it found it unnecessary. Thankfully, it doesn't seem like it had anything to do with the [bug](https://github.com/iovisor/bcc/issues/1621) the commit was trying to resolve, and was just removed cause the author happened to pass by it and figured why not do some cleaning. This means that it is most likely safe to add back in the line that has been removed.

**Comment by jcanseco** [comment]
Please let me know if you approve of the current commits. I'd like to push the next few commits for review when possible. One of these commits include the fix to close perf events.

File: src/cc/frontends/clang/libbpf.h [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+#ifndef LIBBPF_H
+#define LIBBPF_H
+
+#include "compat/linux/bpf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*bpf_create_map_cb_t)(void *cb_cookie, void *data);
```
Comment (joelagnel) [inline src/cc/frontends/clang/libbpf.h]: So this wont fly at all, it was done for prototyping but there's no reason to duplicate entire libbpf.h , upstream wont accept it. Find a way to reuse the existing libbpf.h

File: src/cc/frontends/clang/b_frontend_action.cc [hunk 4]
```diff
@@ -790,8 +790,27 @@ bool BTypeVisitor::VisitVarDecl(VarDecl *Decl) {
...
+        struct bpf_create_map_args args;
+
+        args.type = map_type;
+        args.name = (char *)table.name.c_str();
+        args.key_size = table.key_size;
+        args.value_size = table.leaf_size;
+        args.max_entries = table.max_entries;
+        args.map_flags = table.flags;
+
+        /* TODO: Get return value and store in table.fd */
```
Comment (joelagnel) [inline src/cc/frontends/clang/b_frontend_action.cc]: This is a bit broken and wont fly upstream - did you even try to fix it? 

Can `bpf_create_map_cb` return <= 0 ? If yes, then the below is completely broken because it'll result in a call to `bpf_create_map` for remotes which I don't think makes any sense

File: src/python/bcc/__init__.py [hunk 5]
```diff
@@ -691,12 +805,21 @@ def attach_tracepoint(self, tp=b"", tp_re=b"", fn_name=b""):
...
         if tp_re:
-            for tp in BPF.get_tracepoints(tp_re):
+            for tp in BPF.get_tracepoints(tp_re, bpf_obj=self):
                 self.attach_tracepoint(tp=tp, fn_name=fn_name)
             return
 
         fn = self.load_func(fn_name, BPF.TRACEPOINT)
         (tp_category, tp_name) = tp.split(b':')
+
+        # TODO: Add support for remote cbs
```
Comment (joelagnel) [inline src/python/bcc/__init__.py]: I don't think remote cbs (callbacks) exists anymore so get rid of that

File: src/python/bcc/libbcc.py [hunk 6]
```diff
@@ -17,14 +17,15 @@
...
 lib.bpf_module_create_b.restype = ct.c_void_p
 lib.bpf_module_create_b.argtypes = [ct.c_char_p, ct.c_char_p, ct.c_uint]
 lib.bpf_module_create_c.restype = ct.c_void_p
 lib.bpf_module_create_c.argtypes = [ct.c_char_p, ct.c_uint,
-        ct.POINTER(ct.c_char_p), ct.c_int]
+        ct.POINTER(ct.c_char_p), ct.c_int, _MAP_CB_TYPE, ct.py_object]
 lib.bpf_module_create_c_from_string.restype = ct.c_void_p
 lib.bpf_module_create_c_from_string.argtypes = [ct.c_char_p, ct.c_uint,
-        ct.POINTER(ct.c_char_p), ct.c_int]
+        ct.POINTER(ct.c_char_p), ct.c_int, _MAP_CB_TYPE, ct.py_object]
```
Comment (joelagnel) [inline src/python/bcc/libbcc.py]: Does ctypes need the ct.py_object argument for this?

File: src/python/bcc/remote/adb.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        # adb server is expected to be connected, either
+        # through USB or the network
+        self.client = pe.spawn("adb shell")
+        time.sleep(2)
+
+        if cmd == None:
+            cmd = '/data/bpfd'
+
+        # HACK: Make it work on hikey 4.9 kernel for demo purpose
+        cmd += ' -k 264519'
```
Comment (joelagnel) [inline src/python/bcc/remote/adb.py]: How can this ever go upstream?

File: src/python/bcc/remote/libremote.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        ret = self._remote_send_command(cmd)
+        return ret[0]
+
+    def bpf_attach_kprobe(self, fd, t, evname, fnname):
+        cmd = "BPF_ATTACH_KPROBE {} {} {} {}".format(fd, t, evname, fnname)
+        ret = self._remote_send_command(cmd)
+        return ret[0]
+
+    def bpf_detach_kprobe(self, evname):
+        # this is hard to do since a signal needs to be delivered if
```
Comment (joelagnel) [inline src/python/bcc/remote/libremote.py]: This issue is the same as the perf reader polling issue or it could be the same thing. Either way, why isn't this fixed/cleaned up?

File: tests/test_bcc_remote.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env python
+
+from bcc.remote import shell
+
+s = shell.ShellRemote()
+print s.send_command('testxx')
```
Comment (joelagnel) [inline tests/test_bcc_remote.py]: is this test really expected to print something? And which commit has all the smoke tests your wrote??

**Review by joelagnel — CHANGES_REQUESTED** [review]

File: tests/test_bcc_remote.py (hunk 9 above)
Comment (jcanseco) [reply tests/test_bcc_remote.py]: I intended the tests to come later as a separate commit.

But now that you mention it, it probably is better to include the tests in this commit instead. I'll apply this change shortly.

File: src/python/bcc/__init__.py (hunk 5 above)
Comment (jcanseco) [reply src/python/bcc/__init__.py]: Done.

File: src/python/bcc/remote/adb.py (hunk 7 above)
Comment (jcanseco) [reply src/python/bcc/remote/adb.py]: Done.

File: src/python/bcc/remote/libremote.py (hunk 8 above)
Comment (jcanseco) [reply src/python/bcc/remote/libremote.py]: Done.

File: tests/test_bcc_remote.py (hunk 9 above)
Comment (jcanseco) [reply tests/test_bcc_remote.py]: Done.

File: src/cc/frontends/clang/libbpf.h (hunk 3 above)
Comment (jcanseco) [reply src/cc/frontends/clang/libbpf.h]: Done.

File: src/cc/frontends/clang/b_frontend_action.cc (hunk 4 above)
Comment (jcanseco) [reply src/cc/frontends/clang/b_frontend_action.cc]: Okay I see the problem here. I implemented a fix that dealt with this issue while leveraging the callback mechanism you wrote.

File: src/python/bcc/libbcc.py (hunk 6 above)
Comment (jcanseco) [reply src/python/bcc/libbcc.py]: As far as I can see, yes. This is how we pass around the reference to "self" (where "self" is the BPF object in this case). This allows us to access instance variables and member functions of the BPF object.

The only instance variable we access though is "libremote", so once I apply the commit that turns that into a class variable instead of an instance variable, we can perhaps remove the `ct.py_object` argument and even the `bpf_create_map_cookie` global variable.

**Comment by jcanseco** [comment]
Updates:
- Implemented bpf_close_perf_event_fd() on BPFd side. This is a libbpf.c function that is part of the upstream BCC API changes that rendered the leaked perf reader issue moot.
- Brought in the cmd_parser changes from the BPFd repo. This revamps the input parsing logic and fixes bugs like BPFd's inability to handle commands with < 2 args.
  - Naturally, I also removed all the extra dummy arguments in libremote that were no longer needed
- Addressed all previous PR comments.
  - Note: this includes a fix for the bpf_create_map callback issue where we weren't handling the scenario where invoking bpf_create_map() on remote returns <= 0. 
- Rebased the branch on top of iovisor/master so that we are up-to-date with upstream once again.
- Modified the commit messages to be more coherent (i.e. to have a more "story-like" flow, at least more so than before)

This is all I have for now, I will now proceed to cherry-pick the other bcc-bpfd commits.

File: src/python/bcc/libbcc.py (hunk 6 above)
Comment (jcanseco) [reply src/python/bcc/libbcc.py]: This has been removed in the latest iteration of this PR.

**Comment by jcanseco** [comment]
More updates:
- Added uprobe support, kernel symbol lookup, and userspace symbol lookup
- Changed libremote into a class variable, which resulted in a couple of important clean-ups:
  - It allowed me to remove the redundant functions get_tracepoints_libremote() and get_tracepoints_remote() which originally existed just so we can pass the libremote instance to the static method get_tracepoints()
  - It allowed me to purge bpf_create_map_cb_cookie - a considerable benefit since its usage was quite confusing
  - Simplified the get_kprobe_functions() signature since it no longer needs to be passed the libremote instance

**Comment by jcanseco** [comment]
This PR now effectively contains all commits from bcc-bpfd, except for the ones with PRs still waiting to be approved.

**Comment by jcanseco** [comment]
More updates:
- Cleaned up load_func() to be clearer; it was quite all over the place before
- Removed comments that littered bpfd.c regarding libbpf function signatures
- Added a log() function to libremote so that we're not cluttering BCC with checks for the "BCC_REMOTE_DEBUG" env var

File: src/cc/frontends/clang/CMakeLists.txt (hunk 1 above)
Comment (joelagnel) [reply src/cc/frontends/clang/CMakeLists.txt:13]: Does dropping this commit altogether break anything? Guessing it does

File: src/python/bcc/__init__.py [hunk 10]
```diff
@@ -142,6 +151,8 @@ class BPF(object):
     XDP_PASS = 2
     XDP_TX = 3
 
+    _libremote = None
+
```
Comment (joelagnel) [inline src/python/bcc/__init__.py]: suggest kill newline

File: src/cc/bpf_common.cc [hunk 11]
```diff
@@ -26,7 +28,9 @@ void * bpf_module_create_b(const char *filename, const char *proto_filename, uns
   return mod;
 }
 
-void * bpf_module_create_c(const char *filename, unsigned flags, const char *cflags[], int ncflags) {
+void * bpf_module_create_c(const char *filename, unsigned flags, const char *cflags[], int ncflags, bpf_create_map_cb_t map_cb) {
+  bpf_create_map_cb = map_cb;
```
Comment (joelagnel) [inline src/cc/bpf_common.cc]: Could you add a comment here, why map_cb is needed - either here or in the definition. Its because its not very clear why we're doing this. What do you think

**Review by joelagnel — CHANGES_REQUESTED** [review]
Some comments about the commit messages. Also please mention something about external kernel sources (cross-development) for embedded developers. Check the slides for any other disadvantages worth adding.

> /proc/kallsyms, attach kprobes, etc.).
We're not really directly reading kallsyms anymore, so don't add that here

> For the below, use less "this" and more "it". I learnt early on in my career that using "this" too much doesn't sound right

* This allows us to have kernel sources and the LLVM stack on a separate
host machine (such as the development machine) instead of having to set
these up on the target device, thereby drastically reducing the space
required on a target device to run BCC tools for it. This becomes a
much more critical factor for target devices that have more limited
disk space (e.g. embedded devices). This also allows us to run clang on a
different architecture than the target's architecture, thus facilitating
cross-compilation development.*

For more information, please check out the README in the original
BPFd repository (https://github.com/joelagnel/bpfd) and this LWN article
explaining the purpose of and how BPFd works in more detail
(https://lwn.net/Articles/744522/)

Signed-off-by: Jazel Canseco <jcanseco@google.com>

File: src/cc/frontends/clang/CMakeLists.txt (hunk 1 above)
Comment (jcanseco) [reply src/cc/frontends/clang/CMakeLists.txt:13]: Yes it does. The build fails without it.

File: src/python/bcc/__init__.py (hunk 10 above)
Comment (jcanseco) [reply src/python/bcc/__init__.py]: Done.

File: src/cc/bpf_common.cc (hunk 11 above)
Comment (jcanseco) [reply src/cc/bpf_common.cc]: Done.

**Review by jcanseco — COMMENTED** [review]
> We're not really directly reading kallsyms anymore, so don't add that here

This is not true. We read /proc/kallsyms to perform the kernel symbol lookups.

I also checked the slides, but they mentioned no other disadvantage other than the requirement for having a copy of the kernel sources on the host machine.

**Comment by jcanseco** [comment]
Updates:
- Pulled in the changes from the BCC and BPFd PRs approved yesterday
- Eliminated perf_reader duplication
- Fixed issue regarding BCC's improper exiting of BPFd

**Comment by jcanseco** [comment]
Updates:
- Increased smoke tests and added kernel version checks since we run the tests on a ShellRemote on the host machine (and so the host machine's kernel needs to be at least a certain version in some of the tests)
- Eliminated remote_perf_reader lost_cb
- Removed "-k" flag handling in BPFd since it's no longer needed
- Fix bug where detach_tracepoint() was closing a local perf_event when it should be closing a remote one. This fixes the issue where terminating syscount.py led to crashes.
- Add support for detaching tracepoints on a remote device
- Add support for remote comm_for_pid, fixing the bug wherein syscount.py was reading the comm locally instead of remotely
